File: basic_dsa_four_hundred_fifty_four.java
/**
 * Basic DSA Problems - 4Sum II (LeetCode 454)
 * Count tuples (i,j,k,l) where A[i]+B[j]+C[k]+D[l]=0
 */
public class basic_dsa_four_hundred_fifty_four {

    // Problem: Four arrays. Count a+b+c+d=0.

    /**
     * Map a+b -> count. For each c+d, add count of -(c+d).
     * Time O(n^2), Space O(n^2)
     */
    public static int fourSumCount(int[] nums1, int[] nums2, int[] nums3, int[] nums4) {
        java.util.Map<Integer, Integer> cnt = new java.util.HashMap<>();
        for (int a : nums1)
            for (int b : nums2) cnt.put(a + b, cnt.getOrDefault(a + b, 0) + 1);
        int ans = 0;
        for (int c : nums3)
            for (int d : nums4) ans += cnt.getOrDefault(-(c + d), 0);
        return ans;
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: 4Sum II");
        System.out.println("=".repeat(60));

        System.out.println("A=[1,2] B=[-2,-1] C=[-1,2] D=[0,2] -> " + fourSumCount(new int[]{1,2}, new int[]{-2,-1}, new int[]{-1,2}, new int[]{0,2}));
        System.out.println("A=[0] B=[0] C=[0] D=[0] -> " + fourSumCount(new int[]{0}, new int[]{0}, new int[]{0}, new int[]{0}));
    }
}

File: basic_dsa_four_hundred_fifty_four.cpp
/**
 * Basic DSA - Divide Array Into Equal Pairs (LeetCode 2206)
 * Split nums (length 2n) into n pairs where each pair has two equal numbers.
 */
#include <iostream>
#include <unordered_map>
#include <vector>

bool divideArray(std::vector<int>& nums) {
    std::unordered_map<int, int> cnt;
    for (int x : nums)
        ++cnt[x];
    for (const auto& p : cnt) {
        if (p.second % 2 != 0)
            return false;
    }
    return true;
}

void printVec(const std::vector<int>& v) {
    std::cout << "[";
    for (size_t i = 0; i < v.size(); ++i) {
        if (i)
            std::cout << ", ";
        std::cout << v[i];
    }
    std::cout << "]";
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: Divide Array Into Equal Pairs\n";
    std::cout << std::string(60, '=') << "\n";

    std::vector<int> a1 = {3, 2, 3, 2, 2, 2};
    printVec(a1);
    std::cout << " -> " << (divideArray(a1) ? "true" : "false") << "\n";

    std::vector<int> a2 = {1, 2, 3, 4};
    printVec(a2);
    std::cout << " -> " << (divideArray(a2) ? "true" : "false") << "\n";
}
